packages/structures/src/Structure.ts

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Additional options needed to appropriately construct / patch data
+ *
+ * @internal
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface StructureExtraOptions {}
+
+export const DataTemplatePropertyName = 'DataTemplate';
+export const OptimizeDataPropertyName = '_optimizeData';
+
+/**
+ * Represents a data model from the Discord API
+ *
+ * @privateRemarks
+ * Explanation of the type complexity surround Structure:
+ *
+ * There are two layers of Omitted generics, one here, which allows omitting things at the library level so we do not accidentally
+ * access them, in addition to whatever the user does at the layer above.
+ *
+ * The second layer, in the exported structure is effectively a type cast that allows the getters types to match whatever data template is used
+ *
+ * In order to safely set and access this data, the constructor and patch take data as "partial" and forcibly assigns it to kData. To acommodate this,
+ * kData stores properties as `unknown` when it is omitted, which allows accessing the property in getters even when it may not actually be present.
+ * This is the most technically correct way of representing the value, especially since there is no way to guarantee runtime matches the "type cast."
+ */
+export abstract class Structure<DataType, Omitted extends keyof DataType | '' = ''> {
+	/**
+	 * A construct function used when mixing to allow mixins to set optimized property defaults
+	 *
+	 * @internal
+	 * @remarks This should only be used to set defaults, setting optimized values should be done
+	 * in the mixins `_optimizeData` method, which will be called automatically.
+	 * @param data - The full API data received by the Structure
+	 */
+	protected [kMixinConstruct]?(data: Partial<DataType>): void;
+
+	/**
+	 * The template used for removing data from the raw data stored for each Structure.
+	 *
+	 * @remarks This template should be overriden in all subclasses to provide more accurate type information.
+	 * The template in the base {@link Structure} class will have no effect on most subclasses for this reason.
+	 */
+	protected static DataTemplate: Record<string, unknown> = {};
+
+	/**
+	 * @returns A cloned version of the data template, ready to create a new data object.
+	 */
+	private _getDataTemplate() {
+		return Object.create((this.constructor as typeof Structure).DataTemplate);
+	}
+
+	/**
+	 * The raw data from the API for this struture
+	 *
+	 * @internal
+	 */
+	protected [kData]: Readonly<ReplaceOmittedWithUnknown<Omitted, DataType>>;
+
+	/**
+	 * Creates a new structure to represent API data
+	 *
+	 * @param data - the data from the API that this structure will represent
+	 * @param extraOptions - any additional options needed to appropriately construct a structure from the data
+	 * @remarks To be made public in subclasses
+	 * @internal
+	 */
+	protected constructor(data: Readonly<Partial<DataType>>, ..._rest: unknown[]) {
```

Comment:
Semantically, I think I'd prefer if we went with public here. Protected constructors feel a bit silly when the class is already abstract and requires extending.

```suggestion
	public constructor(data: Readonly<Partial<DataType>>, ..._rest: unknown[]) {
```